Avoid global ordinals in composite aggregation

ywelsch opened the pull request:
Composite aggregations can paginate all buckets from a multi-level aggregation efficiently. It is heavily used by the [*transform*](https://www.elastic.co/guide/en/elasticsearch/reference/current/transforms.html) functionality, for example to convert existing Elasticsearch indices into entity-centric indices that summarize the behavior of users or sessions.

Composite aggregations on `keyword` fields used global ordinals (see ["What are global ordinals"](https://www.elastic.co/guide/en/elasticsearch/reference/master/eager-global-ordinals.html#_what_are_global_ordinals)) to ensure fast comparisons between segments. Global ordinals on high cardinality fields can however use a lot of heap memory as part of the [field data cache](https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-fielddata.html). 

With this PR, composite aggregations no longer need global ordinals, reducing resource consumption for batch-like jobs such as *transform*. The trick is that composite aggregations only need to keep track of the top N composite buckets. Since N is typically small, we can just use the segment ordinal for comparison when collecting inside a segment and remap ordinals when we go to the next segment.

Closes #47452

ywelsch committed:
Avoid loading global ordinals in composite agg

ywelsch committed:
fix

ywelsch committed:
moar fix

ywelsch committed:
nicer

ywelsch committed:
no lookup tracking

ywelsch committed:
cleaner

ywelsch committed:
add breaker

ywelsch committed:
small fix

ywelsch committed:
simpler

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

ywelsch committed:
Merge remote-tracking branch 'elastic/master' into no-glob-ord

not-napoleon commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java:
```diff
@@ -0,0 +1,329 @@
+                    newOrd = newMapping.lookupTerm(newVal);
+                    if (newOrd < 0) {
+                        final BytesRef copiedNewVal = BytesRef.deepCopyOf(newVal);
+                        breakerConsumer.accept(copiedNewVal.bytes.length);
```
Shouldn't we add the bytes to the circuit breaker before making the copy?

not-napoleon commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java:
```diff
@@ -0,0 +1,329 @@
+
+    // is Long.MIN_VALUE when the value represents the missing bucket, or negative if the term is not known to current lookup
+    private Long currentValueOrd;
+    // when term is not known to current lookup, then the term from a previous lookup is stored here, else null
```
I think these comments are a little confusing.  Maybe have one block comment to explain how we're using `Long.MIN_VALUE`, negatives, and nulls here and then annotate the three different pairs of ordinals and values with their purpose?

not-napoleon reviewed (approved):
I left a few notes, but I don't think any of them are blockers on merging.  Thanks for taking this one!

ywelsch commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesCollectorQueue.java:
```diff
@@ -230,13 +232,23 @@ LeafBucketCollector getLeafCollector(Comparable forceLeadSourceValue,
-            collector = arrays[last].getLeafCollector(context, collector);
+            collector = valuesSource.getLeafCollector(context, collector);
+        }
+        if (requiresRehashing) {
```
Correct. By remapping the ordinals, the keys in this map are mutable, which means that the previously stored entries can't be found anymore with this key. The hashCode of `Slot` is based on the ordinal number (see `OrdinalValuesSource.hashCode()`), which is only meaningful during the processing of a single leaf. When switching leaves, ordinals are remapped, and the keys in the `CompositeValuesCollectorQueue.map` (the `Slot`s) have therefore their hashCode changed under them (which means that looking up these slots in the map won't yield the correct entry anymore). Rebuilding the map when switching leaves (aka remapping ordinals) ensures that the hashCode used to insert the entry in the map corresponds again to the current hashCode of the key.

ywelsch committed:
Merge remote-tracking branch 'elastic/master' into no-glob-ord

ywelsch committed:
review feedback (more comments) + clearer invariants

ywelsch committed:
Merge remote-tracking branch 'elastic/master' into no-glob-ord

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java:
```diff
@@ -0,0 +1,372 @@
+        }
+        BytesRef term = (BytesRef) value;
+        final SortedSetDocValues dvs = docValuesFunc.apply(context);
+        remapOrdinals(lookup, dvs);
```
We call this function once per bucket in SortedDocsProducer so we don't need to remap the ordinals each time.
We should keep track of the context ord and remap only if the provided context ord is different ?

ywelsch replied:
good idea. addressed in 64b1872

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeValuesCollectorQueue.java:
```diff
@@ -230,13 +232,23 @@ LeafBucketCollector getLeafCollector(Comparable forceLeadSourceValue,
                                          LeafReaderContext context, LeafBucketCollector in) throws IOException {
         int last = arrays.length - 1;
         LeafBucketCollector collector = in;
+        boolean requiresRehashing = false;
```
We call this function multiple times per leaf in the sorted case. So the remapping is not always necessary. We should compare the context.ord with the previous one to avoid doing the work when it's not needed.

ywelsch replied:
good idea. addressed in 64b1872

jimczi reviewed (commented):
The logic looks good to me. I left two comments to avoid the remapping when it's not needed.

ywelsch committed:
Avoid remapping when leafreader unchanged

ywelsch committed:
Merge remote-tracking branch 'elastic/master' into no-glob-ord

jimczi reviewed (approved):
LGTM, thanks!

ywelsch committed:
avoid too much slowdown in tests by checking class invariant so often

CompositeValuesCollectorQueueTests.testRandom was timing out as it took too long to run,
because we kept checking the invariant every time on copyCurrent

ywelsch committed:
woops, invariant in the wrong place

ywelsch committed:
Merge remote-tracking branch 'elastic/master' into no-glob-ord
